File: frontend/nextjs-app/src/lib/services/stories-service.ts
import { axiosClient } from '../api/axios-client';
import type {
  Story,
  StoryWithDetails,
  StorySearchResponse,
  Character,
  LessonInStory,
  StorySource,
  StoryCategory,
  PaginationOptions,
  StoryFilters,
  CharacterFilterOptions,
  ThemeFilterOptions,
  CharacterFilters,
} from '@/types/stories';

/**
 * Service for managing Islamic stories
 */
class StoriesService {
  private readonly baseUrl = '/api/stories';

  /**
   * Get stories by category
   */
  async getStoriesByCategory(
    category: StoryCategory,
    options?: PaginationOptions
  ): Promise<Story[]> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/category/${category}`,
        {
          params: {
            limit: options?.limit,
            offset: options?.offset,
          },
        }
      );
      return response.data.data;
    } catch (error) {
      console.error('Error fetching stories by category:', error);
      throw error;
    }
  }

  /**
   * Get a single story by ID
   */
  async getStory(
    storyId: string,
    includeDetails: boolean = true
  ): Promise<StoryWithDetails> {
    try {
      const response = await axiosClient.get(`${this.baseUrl}/stories/${storyId}`, {
        params: {
          include_details: includeDetails,
        },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error fetching story:', error);
      throw error;
    }
  }

  /**
   * Search stories
   */
  async searchStories(
    query: string,
    filters?: StoryFilters,
    options?: PaginationOptions
  ): Promise<StorySearchResponse> {
    try {
      const response = await axiosClient.get(`${this.baseUrl}/stories`, {
        params: {
          query,
          categories: filters?.categories?.join(','),
          age_groups: filters?.ageGroups?.join(','),
          authenticity_levels: filters?.authenticityLevels?.join(','),
          time_periods: filters?.timePeriods?.join(','),
          themes: filters?.themes?.join(','),
          limit: options?.limit,
          offset: options?.offset,
        },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error searching stories:', error);
      throw error;
    }
  }

  /**
   * Get stories by character
   */
  async getStoriesByCharacter(
    characterName: string,
    options?: CharacterFilterOptions
  ): Promise<Story[]> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/character/${encodeURIComponent(characterName)}`,
        {
          params: {
            character_type: options?.characterType,
            include_related: options?.includeRelated,
            limit: options?.limit,
            offset: options?.offset,
          },
        }
      );
      return response.data.data.stories.map((item: any) => item.story);
    } catch (error) {
      console.error('Error fetching stories by character:', error);
      throw error;
    }
  }

  /**
   * Get stories by theme
   */
  async getStoriesByTheme(
    theme: string,
    options?: ThemeFilterOptions
  ): Promise<Story[]> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/theme/${encodeURIComponent(theme)}`,
        {
          params: {
            lesson_type: options?.lessonType,
            moral_category: options?.moralCategory,
            age_group: options?.ageGroup,
            limit: options?.limit,
            offset: options?.offset,
          },
        }
      );
      return response.data.data.stories.map((item: any) => item.story);
    } catch (error) {
      console.error('Error fetching stories by theme:', error);
      throw error;
    }
  }

  /**
   * Get story lessons
   */
  async getStoryLessons(storyId: string): Promise<LessonInStory[]> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/stories/${storyId}/lessons`
      );
      return response.data.data;
    } catch (error) {
      console.error('Error fetching story lessons:', error);
      throw error;
    }
  }

  /**
   * Get story sources
   */
  async getStorySources(storyId: string): Promise<StorySource[]> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/stories/${storyId}/sources`
      );
      return response.data.data;
    } catch (error) {
      console.error('Error fetching story sources:', error);
      throw error;
    }
  }

  /**
   * Search characters
   */
  async searchCharacters(
    query: string,
    filters?: CharacterFilters
  ): Promise<Character[]> {
    try {
      const response = await axiosClient.get(`${this.baseUrl}/characters/search`, {
        params: {
          query,
          character_type: filters?.characterType,
          historical_period: filters?.historicalPeriod,
          limit: filters?.limit,
          offset: filters?.offset,
        },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error searching characters:', error);
      throw error;
    }
  }

  /**
   * Get category statistics
   */
  async getCategoryStatistics(): Promise<Record<string, number>> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/analytics/categories`
      );
      return response.data.data;
    } catch (error) {
      console.error('Error fetching category statistics:', error);
      throw error;
    }
  }

  /**
   * Verify story integrity
   */
  async verifyStoryIntegrity(storyId: string): Promise<boolean> {
    try {
      const response = await axiosClient.get(
        `${this.baseUrl}/stories/${storyId}/integrity`
      );
      return response.data.data.is_valid;
    } catch (error) {
      console.error('Error verifying story integrity:', error);
      throw error;
    }
  }
}

export const storiesService = new StoriesService();

File: frontend/mobile/lib/core/services/stories_service.dart
import 'package:dio/dio.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';
import '../../features/stories/data/models/story_model.dart';

/// Service for managing Islamic stories
class StoriesService {
  final DioClient _dioClient;

  StoriesService(this._dioClient);

  /// Get stories by category
  Future<List<StoryModel>> getStoriesByCategory(
    StoryCategory category, {
    int? limit,
    int? offset,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/category/${category.name}',
        queryParameters: {
          if (limit != null) 'limit': limit,
          if (offset != null) 'offset': offset,
        },
      );

      final data = response.data['data'] as List;
      return data.map((json) => StoryModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to get stories by category: $e');
    }
  }

  /// Get a single story by ID
  Future<StoryWithDetailsModel> getStory(
    String storyId, {
    bool includeDetails = true,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/$storyId',
        queryParameters: {
          'include_details': includeDetails,
        },
      );

      return StoryWithDetailsModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Failed to get story: $e');
    }
  }

  /// Search stories
  Future<List<StoryModel>> searchStories(
    String query, {
    List<StoryCategory>? categories,
    List<AgeGroup>? ageGroups,
    List<AuthenticityLevel>? authenticityLevels,
    int? limit,
    int? offset,
  }) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.stories,
        queryParameters: {
          'query': query,
          if (categories != null)
            'categories': categories.map((c) => c.name).join(','),
          if (ageGroups != null)
            'age_groups': ageGroups.map((a) => a.name).join(','),
          if (authenticityLevels != null)
            'authenticity_levels':
                authenticityLevels.map((a) => a.name).join(','),
          if (limit != null) 'limit': limit,
          if (offset != null) 'offset': offset,
        },
      );

      final results = response.data['data']['results'] as List;
      return results.map((json) => StoryModel.fromJson(json['story'])).toList();
    } catch (e) {
      throw Exception('Failed to search stories: $e');
    }
  }

  /// Get stories by character
  Future<List<StoryModel>> getStoriesByCharacter(
    String characterName, {
    CharacterType? characterType,
    int? limit,
    int? offset,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/character/$characterName',
        queryParameters: {
          if (characterType != null) 'character_type': characterType.name,
          if (limit != null) 'limit': limit,
          if (offset != null) 'offset': offset,
        },
      );

      final stories = response.data['data']['stories'] as List;
      return stories
          .map((json) => StoryModel.fromJson(json['story']))
          .toList();
    } catch (e) {
      throw Exception('Failed to get stories by character: $e');
    }
  }

  /// Get stories by theme
  Future<List<StoryModel>> getStoriesByTheme(
    String theme, {
    LessonType? lessonType,
    MoralCategory? moralCategory,
    AgeGroup? ageGroup,
    int? limit,
    int? offset,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/theme/$theme',
        queryParameters: {
          if (lessonType != null) 'lesson_type': lessonType.name,
          if (moralCategory != null) 'moral_category': moralCategory.name,
          if (ageGroup != null) 'age_group': ageGroup.name,
          if (limit != null) 'limit': limit,
          if (offset != null) 'offset': offset,
        },
      );

      final stories = response.data['data']['stories'] as List;
      return stories
          .map((json) => StoryModel.fromJson(json['story']))
          .toList();
    } catch (e) {
      throw Exception('Failed to get stories by theme: $e');
    }
  }

  /// Get story lessons
  Future<List<LessonInStoryModel>> getStoryLessons(String storyId) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/$storyId/lessons',
      );

      final lessons = response.data['data'] as List;
      return lessons.map((json) => LessonInStoryModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to get story lessons: $e');
    }
  }

  /// Get story sources
  Future<List<StorySourceModel>> getStorySources(String storyId) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/$storyId/sources',
      );

      final sources = response.data['data'] as List;
      return sources.map((json) => StorySourceModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to get story sources: $e');
    }
  }

  /// Get characters
  Future<List<CharacterModel>> searchCharacters(
    String query, {
    CharacterType? characterType,
    TimePeriod? historicalPeriod,
    int? limit,
    int? offset,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/characters/search',
        queryParameters: {
          'query': query,
          if (characterType != null) 'character_type': characterType.name,
          if (historicalPeriod != null)
            'historical_period': historicalPeriod.name,
          if (limit != null) 'limit': limit,
          if (offset != null) 'offset': offset,
        },
      );

      final characters = response.data['data'] as List;
      return characters.map((json) => CharacterModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to search characters: $e');
    }
  }

  /// Get category statistics
  Future<Map<String, int>> getCategoryStatistics() async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/analytics/categories',
      );

      final stats = response.data['data'] as Map<String, dynamic>;
      return stats.map((key, value) => MapEntry(key, value as int));
    } catch (e) {
      throw Exception('Failed to get category statistics: $e');
    }
  }

  /// Verify story integrity
  Future<bool> verifyStoryIntegrity(String storyId) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.stories}/$storyId/integrity',
      );

      return response.data['data']['is_valid'] as bool;
    } catch (e) {
      throw Exception('Failed to verify story integrity: $e');
    }
  }
}
